Userland/xargs.cpp

```diff
@@ -0,0 +1,186 @@
+    args_parser.add_option(specified_delimiter, "Split the input items with the specified character", "delimiter", 'd', "delim");
+    args_parser.add_option(verbose, "Display each command before executing it", "verbose", 'v');
+    args_parser.add_option(file_to_read, "Read arguments from the specified file instead of stdin", "arg-file", 'a', "file");
+    args_parser.parse(argc, argv);
+
+    char entry_separator = split_with_nulls ? '\0' : *specified_delimiter;
+
+    StringView replace_view { replace_str };
+
+    Vector<Vector<StringView>> argument_parts;
+    argument_parts.ensure_capacity(arguments.size());
+    size_t index = 0;
+    bool some_argument_has_replacement = false;
+
+    for (auto argument : arguments) {
+        StringView arg { argument };
+        Vector<StringView> parts;
+        size_t previous_index = 0;
+        if (auto first_index = arg.find_first_of(replace_view); first_index.has_value()) {
+            some_argument_has_replacement = true;
+            while (first_index.has_value()) {
+                auto current_index = first_index.value();
+                auto substring_until_replacement = arg.substring_view(previous_index, current_index);
+                parts.append(substring_until_replacement);
+                previous_index = current_index;
+                first_index = arg.substring_view(current_index + replace_view.length(), arg.length() - current_index - replace_view.length()).find_first_of(replace_view);
+            }
+            parts.append(arg.substring_view(previous_index + replace_view.length(), arg.length() - previous_index - replace_view.length()));
+        } else {
+            parts.append(arg);
+        }
+        argument_parts.append(move(parts));
+        ++index;
+    }
+
+    // Append a '{}' at the end if no arguments have replacements.
+    if (!some_argument_has_replacement) {
+        Vector<StringView> parts;
+        parts.append("");
+        parts.append("");
+        argument_parts.append(move(parts));
+    }
+
+    FILE* fp = stdin;
+    bool is_stdin = true;
+
+    if (StringView { "-" } != file_to_read) {
+        // A file was specified, try to open it.
+        fp = fopen(file_to_read, "r");
+        if (!fp) {
+            perror("fopen");
+            return 1;
+        }
+        is_stdin = false;
+    }
+
+    char* line = nullptr;
+    size_t buffer_size = 0;
+
+    StringBuilder builder;
+    Vector<char*> child_argv;
+
+    child_argv.resize(argument_parts.size() + 2);
+    child_argv[0] = const_cast<char*>(command);
+    child_argv[child_argv.size() - 1] = 0;
+
+    int devnull_fd = 0;
+
+    if (is_stdin) {
+        devnull_fd = open("/dev/null", O_RDONLY);
+        if (devnull_fd < 0) {
+            perror("open");
+            return 1;
+        }
+    }
+
+    for (;;) {
+        auto line_size = getdelim(&line, &buffer_size, entry_separator, fp);
```

Comment:
As far as I can see, this runs the command with a *single* argument out of those specified in the input file. But that's not what `xargs` does, it passes them all to one invocation (unless they don't fits into argv). E.g.

```
$ xargs ps <<EOF
35
42
411
EOF
```

should invoke

```
$ ps 35 42 411
```

not

```
$ ps 35
$ ps 42
$ ps 411
```

Reply:
I _did_ document that behaviour in the manpage, since I didn't add `-L`, I let the default behaviour be `-L 1`

I personally like this better, but I guess I can change it to work more like posix xargs

Reply:
> since I didn't add `-L`, I let the default behaviour be `-L 1`

`L` is (I bet) for "limit". It's for limiting how many items you process in one go; the default is no limit at all, not one.

I mean, it's fine to not support advanced options / modes & stuff, but `xargs`, even in its simplest form, does a different thing from what you've implemented.

Side note: for the longest time, I actually thought `xargs` does just that, invokes one command per item.

Reply:
Oh, I _knowingly_ didn't do what xargs does, I prefer it with this behaviour.
but again, I can make it cram as many arguments as it can into a single command...

Reply:
_do_ we want to do it the posix way?
it's a bit of a rewrite if we do

Reply:
I'm afraid we do

Reply:
T-T
Now i have to support half a million nonsensical modes just to appease Dr. POSIX.

Reply:
How about we drop `-I` support instead? You'd simply have to append items to the argv.

Reply:
Let's just full-arse this thing lol.
I _want_ the one-line-per-command functionality.

Reply:
`-L` support full-arsed :^)